### DISCUSSION TITLE
Implements JobClientFilter with no success

### DISCUSSION THREAD

**Discussion by zMcKracken** [jobrunr/jobrunr#1105] [Q&A] [answered]
Hello everyone,
I'm using jobrunr 7.2.3 with springboot 3.3.x.
I'd like to intercept when a job change state and based on some criteria eliminate it, re-queue it or just let it go.
Because I'm using the free version I use Fluent API to register the filter like this:

```
@Configuration
public class JobRunnerConfiguration {

    @Bean
    public JobRunrConfiguration.JobRunrConfigurationResult jobRunrConfiguration() {
        return JobRunr.configure()
                .useStorageProvider(new DefaultSqlStorageProvider(dataSource, new PostgresDialect(), StorageProviderUtils.DatabaseOptions.SKIP_CREATE))
                .withJobFilter(new CustomJobFilter())
                .useBackgroundJobServer(1)
                .useDashboard()
                .initialize();
    }
}
```

CustomJobFilter is a class I take as example from the documentation:

```
public class CustomJobFilter implements JobClientFilter, JobServerFilter {

    public CustomJobFilter() {
       logger.info("initialized");
    }

    public void onCreating(AbstractJob job) {
        logger.info( "onCreating");
    }

    public void onCreated(AbstractJob job) {
            logger.info( "created");
    }

    public void onProcessingSucceeded(Job job) {
            logger.info("onProcessingSucceeded.");
    }

    public void onProcessingFailed(Job job, Exception e) {
            logger.info("onProcessingFailed.");
    }
}
```

I can intercept filter from JobServerFilter so as example onProcessingSucceeded works fine but onCreating does not work at all.
Did I miss something?
If I want to intercept before the job goes to state Processing,as example, for deleting it, is it correct to use the ElectStateFilter?

thanks a lot for any advice!

**Comment by auloin** [comment]
Hi @zMcKracken, can you share a reproducing repository for the issue you're having?

> If I want to intercept before the job goes to state Processing,as example, for deleting it, is it correct to use the ElectStateFilter?

Yes, please refer to `ElectStateFilter`'s Javadoc for more details.

**Reply by zMcKracken** [reply]
Hi @auloin ,
I create a small demo project you can download it here: [https://github.com/zMcKracken/demojbr](https://github.com/zMcKracken/demojbr) 
You can launch a job doing a GET to this address: http://localhost:9070/job-runner/api/v1/test/jobI 

thank you very much

**Reply by auloin** [reply]
@zMcKracken it's as @rdehuyss says, you're mixing up the spring-boot-starter and the fluent API. I think you should stick to one approach.

**Reply by zMcKracken** [reply]
Hi @auloin,
Now I see what you mean, my apologize. I was creating BackgroundJobServer using Fluent API but inside a bean in a @Configuration class so using SpringBoot Framework. As a test I create and send a message inside my main class and indeed now it works as expected. 
But I need to use it inside spring framework so I stick with the jobrunr-spring-boot-3-starter, which is fine but I don't know how to set the customFilter in this configuration, any idea?

thanks again and sorry for the misunderstanding

**Reply by zMcKracken** [reply]
@auloin @rdehuyss  I sort it out, thanks for your support and to point me in the right direction

**Reply by rdehuyss** [reply]
Happy to hear you resolved it!

**Comment by rdehuyss** [comment, marked as answer]
I'm afraid your mixing Spring beans with the fluent API which will not work. The code you added just starts a second BackgroundJob server and that may indeed use the filter. But as you have 2 servers, it may be indeterministic.

Have a look at the Spring Boot Autoconfiguration here in JobRunr on how to solve this.

**Reply by zMcKracken** [reply]
Hi @rdehuyss,
thanks a lot for the insight, indeed I messed up a bit. Turns out I already have the jobrunr-spring-boot-3-starter in my pom and I'm using the properties for the configuration, I completely forgot that part, my apologize.
I don't see in properties a way to add a custom filter, should I remove the springboot start from the pom and leave the creation of the BackgroundJob server as in my original post using Fluent API, or there a more elegant way to add a filter keeping the springboot starter, considering I'm in a free version now.

thanks a lot!

UPDATE:
I make some test first I remove the creation via Fluent API and leave the springboot-starter with these properties:

org.jobrunr.background-job-server.enabled: true
org.jobrunr.background-job-server.worker-count: 1
org.jobrunr.dashboard.enabled: false

Everything works fine (no custom filter added), I see in the backgroundjobservers table the creation of a record . Then I put false in the job-server.enabled and no record in the table was created, as expected. Then I create the server via fluent API same code as before, keeping false (or commented) the server.enabled properties and as expected the server is created, just one.
The custom filter is added but still onCreating or onCreated is not called when creating a job, here the code that I forget to paste that I use to submit a job:
```

 jobScheduler.create(JobBuilder.aJob()
                .withName("A job requested ")
                .withLabels("acme")
                .withDetails(() -> defaultJobRunnerService.cacl("some parameteres")));
```
